software/AEB.md

```diff
@@ -8,30 +8,44 @@ is non-deterministic, and thus requires additional deterministic safety systems
 
 ## Inputs
 - **Perception CAN**
-  - **Range sensors** - any sensor that outputs `Range Data` frames.
+  - **UART to LD06 LiDAR** - Tiny 2d LiDAR for perception.
   - **Encoder data** - any sensor that outputs `Encoder Count` frames.
 
 ## Outputs 
 - **EStop** - direct connection to Estop circuit to allow for emergency braking.
 
 ## Algorithm:
+The general idea will be to take all Lidar points, constrain them to an angle range, then map those 
+points onto a bounded voxel grid, where each voxel is either occupied by an obstacle, free space, or will be occupied by the kart. 
+we will then use the ackermann equations to determine the current path
+of the kart, and create an interpolated line along the path of the kart with the width of the kart.
+If the kart's path attempts to set a voxel high that was already set high by the point clouds, then we have a collision.
+Because this grid should be about ourselves, we can take the euclidean distance from that collision to the kart,
+and find our time to collision (TTC) by finding our velocity at that point along the curve. If that TTC is too low, then estop.
 
-```mermaid
-graph TD
-    A[init] --parallel--> wait[Wait for CAN range message]
-    wait --> B{TTC under n seconds?}
-    A --parallel--> vel[Wait for CAN velocity message]
-        vel --> velrec[Update velocity variable]
-        velrec --> vel
-
-    B -- yes --> E[Trigger estop]
-    B -- no --> wait
-```
+- Create an `nxn` grid, where n is the number of voxels, and voxels are some real size measurement `m`.
+- For each LiDAR point in the scan `p`
+  - Get the x,y coordinates of the point relative to the kart by converting from polar to euclidean
+  - If that x,y point is in an unoccupied cell, then mark it as occupied by obstacle (TODO is this trival?)
+- Use the configured wheelbase, steering angle, ect. to create a function for the arc created by the current motion of the kart, herby referred to as `f(t)`
+- For each `t`; `0 < t <= some max timestep`
+  - `kart_x, kart_y = f(t)`
+  - Mark that cell containing `kart_x, kart_y` as occupied by kart, call that cell `C`
```

Comment:
Marking the cell will depend on the size of 'm'. If m is small, then we would need to mark several cells depending on size of kart

Reply:
Another note, the x and y returned by the function will be the x and y position of the center of the rear axle of the kart

Reply:
> Marking the cell will depend on the size of 'm'. If m is small, then we would need to mark several cells depending on size of kart

This is covered by the line below this

Reply:
> Another note, the x and y returned by the function will be the x and y position of the center of the rear axle of the kart

This shouldn't matter I dont think, as we will iteratively fill the voxels with the wirdth of the kart, which will include front to back when done.